Packs/PrismaSASE/Integrations/PrismaSASE/PrismaSASE.py

```diff
@@ -0,0 +1,819 @@
+    "description": "",
+    "log_setting": "",
+    "application": [],
+    "category": [],
+    "destination": [],
+    "destination_hip": [],
+    "profile_setting": {},
+    "service": [],
+    "source": [],
+    "source_hip": [],
+    "source_user": [],
+    "tag": [],
+    "from": [],
+    "to": [],
+    "disabled": "",
+    "negate_source": "",
+    "negate_destination": ""
+}
+
+
+class Client(BaseClient):
+    """Client class to interact with the service API
+
+    This Client implements API calls to the Saas Security platform, and does not contain any XSOAR logic.
+    Handles the token retrieval.
+
+    :param base_url (str): Saas Security server url.
+    :param client_id (str): client ID.
+    :param client_secret (str): client secret.
+    :param verify (bool): specifies whether to verify the SSL certificate or not.
+    :param proxy (bool): specifies if to use XSOAR proxy settings.
+    """
+
+    def __init__(self, base_url: str, client_id: str, client_secret: str, oauth_url: str, verify: bool, proxy: bool, **kwargs):
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.oauth_url = oauth_url
+
+        super().__init__(base_url=base_url, verify=verify, proxy=proxy, **kwargs)
+
+    @staticmethod
+    def build_security_rule(args: dict):
+
+        # This method will combine the args into a JSON formatted request body
+        rule = {}
+        keys = args.keys()
+        for key in SECURITYRULE_FIELDS:
+            if key in keys:
+
+                if args.get(key) != "" and args.get(key) != "None":
+                    if key == 'profile_setting':
+                        if isinstance(args.get(key), list):
+                            val = args.get(key)
+                        else:
+                            val = [x.strip() for x in args.get(key).split(',')]   # type: ignore
+
+                        rule[key] = {'group': val}
+
+                    elif isinstance(SECURITYRULE_FIELDS.get(key), str):
+                        rule[key] = args.get(key)   # type: ignore
+
+                    elif isinstance(SECURITYRULE_FIELDS.get(key), list):
+                        if isinstance(args.get(key), list):
+                            val = args.get(key)
+                        else:
+                            val = [x.strip() for x in args.get(key).split(';')]   # type: ignore
+                        rule[key] = val   # type: ignore
+
+        return rule
+
+    def create_security_rule(self, rule: dict, folder: str, position: str, tsg_id: str):
+        uri = '/sse/config/v1/security-rules'
+        access_token = self.get_access_token(tsg_id)
+
+        query_params = {
+            'folder': encode_string_results(folder),
+            'position': encode_string_results(position)
+        }
+
+        headers = {
+            'Accept': 'application/json',
+            'Content-Type': 'application/json',
+            'Authorization': f"Bearer {access_token}"
+        }
```

Comment:
Hey Shachar, the SASE API is multi-tenant capable and the tenant structure is hierarchical.  The TSG (tenant services group) is an identifier for a particular tenant.  A single API service account can have access to the root tenant and any number of sub-tenants underneath.  The scope / target of the API call is determined via the requested authorization token.   My intent was to enable someone to configure a single instance of the integration, but allow them to target any sub-tenant that the service account may have access to.  I really just followed what I saw in other token based authorizations but added the TSG variable in order to store multiple tokens.

Reply:
I'm not sure I understand your suggestion on self._headers.

Reply:
Hi - thanks for your detailed explanation :) 
About the multi-tenant, can you estimate how many tenants will be used in parallel for the average user?

Reply:
For an SP, it could be hundreds (the platform limits mention 100 Tenants per immediate parent but can also be 10 levels deep with a max of 1000 listed for the hierarchy.  These limits are not hard limits though as far as I understand.)

For a normal enterprise or mid sized customer, probably most would be under 5.

Reply:
OK thank you :) 
So I think we will keep your implementation that allows running each command on a different tenant by generating a separate access token for each tenant and storing it in the integration context. 
 I checked if we have any size limit/performance limitations for what can be stored in the integration's context, and it seems there isn't a technical limit (it can cause performance issues if the object is big (MBs) but that is not the case).